### PR TITLE
Windows,Bazel client: check embedded tools faster

### PR DESCRIPTION
**Pull request by laszlocsomor**
The Bazel client on Windows is now 50% faster to
check the embedded tools than it was before.

Results:
- Linux: 20 ms -> 6 ms
- Windows: 294 ms -> 133 ms

Measurements were done with n=10 runs and a hot
server, using blaze::GetMillisecondsMonotonic().

Previously the client performed the same tasks
multiple times while trying to determine if a path
was a good extracted binary. (E.g. converted the
path to Windows format multiple times, checked if
it was a directory twice, opened the path twice.)

Now the client performes these tasks only once,
e.g. it converts path once and stats only once.

See https://github.com/bazelbuild/bazel/issues/5444

### CODE REVIEW

File: src/main/cpp/blaze.cc [hunk 1]
```diff
@@ -1062,27 +1062,14 @@ static void ExtractData(const string &self_path) {
...
-            << "corrupt installation: file '" << path
-            << "' missing. Please remove '" << globals->options->install_base
-            << "' and try again.";
-      }
-      // Check that the timestamp is in the future. A past timestamp would
-      // indicate that the file has been tampered with.
-      // See ActuallyExtractData().
-      bool is_in_future = false;
-      if (!mtime.get()->GetIfInDistantFuture(path, &is_in_future)) {
+      bool is_good = false;
```
Comment (cvcal) [inline src/main/cpp/blaze.cc]: I don't see the point of having both a success-bool return value and this additional boolean marker. The errors are also almost identical.

File: src/main/cpp/util/file_platform.h [hunk 2]
```diff
@@ -31,10 +31,12 @@ class IFileMtime {
...
-  // Queries the mtime of `path` to see whether it's in the distant future.
-  // Returns true if querying succeeded and stores the result in `result`.
-  // Returns false if querying failed.
-  virtual bool GetIfInDistantFuture(const std::string &path, bool *result) = 0;
+  // Checks if `path` is a good "extracted binary", i.e. it is either an
+  // embedded binary that Bazel extracted from a previous run, and it hasn't
+  // been tampered with (its mtime is unchanged), or it's a directory.
+  // Returns false if querying the information failed, either due to I/O error
+  // or due to the `path` missing or it being a broken symlink.
+  virtual bool CheckExtractedBinary(const std::string &path, bool *result) = 0;
```
Comment (cvcal) [inline src/main/cpp/util/file_platform.h]: I would personally put this in blaze_util_platform, if the comment is true and that its properties are quite Bazel-specific

**Comment by cvcal** [comment]
Did you also time the difference on linux?

File: src/main/cpp/util/file_platform.h (hunk 2 above)
Comment (laszlocsomor) [reply src/main/cpp/util/file_platform.h]: I agree, but moving it there would make this PR a lot more complicated. Let me not do that now.

File: src/main/cpp/blaze.cc (hunk 1 above)
Comment (laszlocsomor) [reply src/main/cpp/blaze.cc]: Done.

**Comment by laszlocsomor** [comment]
> Did you also time the difference on linux?

I just did. The PR improves that too. Updated the commit message.

**Comment by cvcal** [comment]
Assuming this is superseded with https://github.com/bazelbuild/bazel/pull/5445

**Comment by laszlocsomor** [comment]
> Assuming this is superseded with #5448

No, they are separate PRs and will be imported as separate commits.
However GitHub seems to have no good support for something like a Diffbase -- GitHub shows the entire branch as it diverged from `master`, and I can't tell it to display it based on this PR (#5445).

**Pull request closed without merging**